### PR TITLE
[drop counters] Adds initial support for testing configurable drop counters

### PR DESCRIPTION
**Pull request by daall**
- Defines common drop counter methods
- Defines utilities for mocking servers for testing
- Adds L3_EGRESS_LINK_DOWN test case

Signed-off-by: Danny Allen <daall@microsoft.com>

### Description of PR

Summary: Adds initial support for testing configurable drop counters (e.g. common utilities for other test cases to use).

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### How did you do it?
I added some utility functions to interact with drop counters. I also added some more general utilities for mocking servers from PTF, which several of the drop counter scenarios rely on for accurate forwarding behavior.

In order to verify these additions I added an initial test case to exercise the utilities, shown below:

#### How did you verify/test it?
```
daall@00c187053803:/var/src/Networking-acs-sonic-mgmt/tests$ pytest drop_counters/test_config_counters.py --testbed=vms11-t0-on-4 --inventory=../ansible/str --testbed_file=../ansible/testbed.csv --host-pattern=vms11-t0-on-4 --module-path=../ansible/library --disable_loganalyzer --show-capture=no -v
=============================================================================================== test session starts ================================================================================================
platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python
cachedir: .pytest_cache
ansible: 2.8.7
rootdir: /var/src/Networking-acs-sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.1.3, xdist-1.28.0, ansible-2.2.2, repeat-0.8.0
collected 2 items                                                                                                                                                                                                  

drop_counters/test_config_counters.py::test_neighbor_link_down[PORT_INGRESS_DROPS-L3_EGRESS_LINK_DOWN] PASSED                                                                                                [ 50%]
drop_counters/test_config_counters.py::test_neighbor_link_down[SWITCH_INGRESS_DROPS-L3_EGRESS_LINK_DOWN] SKIPPED                                                                                             [100%]

================================================================================================= warnings summary =================================================================================================
drop_counters/test_config_counters.py::test_neighbor_link_down[PORT_INGRESS_DROPS-L3_EGRESS_LINK_DOWN]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible/module_dispatcher/v28.py:70: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============================================================================================= short test summary info ==============================================================================================
SKIPPED [1] /var/src/Networking-acs-sonic-mgmt/tests/drop_counters/test_config_counters.py:75: Counter type not supported on target DUT
================================================================================= 1 passed, 1 skipped, 1 warnings in 80.03 seconds =================================================================================
```
#### Any platform specific information?
Not that I'm aware of. The tests should use the query functionality to handle different platforms.

#### Supported testbed topology if it's a new test case?
t0 only, at this stage. Once we have the tests stabilized with the t0 topology we will look to enable it on more topologies.

### CODE REVIEW

File: tests/drop_counters/configurable_drop_counters.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    counters = {}
+    reasons = {}
+
+    # Delete the header.
+    output = output["stdout_lines"][2:]
+
+    # Extract the supported counter types and quantities from the first
+    # part of the output. We'll parse until we hit the empty line.
+    line = output.pop(0)
+    while line:
```
Comment (yxieca) [inline tests/drop_counters/configurable_drop_counters.py:62]: You can use while len(output): to loop, then you can move the output.pop(0) into the while loop. Save one duplicate line.

Comment (yxieca) [reply tests/drop_counters/configurable_drop_counters.py:62]: I see, you are stopping at empty line. never mind.

File: tests/drop_counters/configurable_drop_counters.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        reasons[line[0]] = []
+        line = output.pop(0)
+
+    # Extract the supported drop reasons from the second part of the output.
+    curr_type = []
+    for line in output:
+        if not line:
+            continue
+        elif line in SUPPORTED_COUNTER_TYPES:
+            curr_type = reasons[line]
```
Comment (yxieca) [inline tests/drop_counters/configurable_drop_counters.py:75]: what if line is not in reasons?

File: tests/drop_counters/configurable_drop_counters.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    status_columns = _PARSER_PARAMETERS[counter_type].status_columns
+
+    # Skip any leading rows in the output
+    trimmed_output = counts_output[leading_rows:]
+
+    # Skip over the status columns to get the counter names
+    counter_names = trimmed_output[0].split()[status_columns:]
+
+    # Skip over the header rows to get to the actual counts
+    interface_counts = trimmed_output[header_rows:]
```
Comment (yxieca) [inline tests/drop_counters/configurable_drop_counters.py:148]: Just curious, why did you have to trim 3 times instead of just add up 3 overheads and skip once?

File: tests/drop_counters/mock_server.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        Initializes the server.
+
+        Args:
+            server_addr (IPAddress): The IP address of the server.
+            server_prefix (int): The network prefix of the server.
+            neighbor_addr (IPAddress): The IP address of the device the server is connected to
+                (typically an IP interface in SONiC).
+            neighbor_iface (str): The interface associated with the `neighbor_addr`.
+            neighbor_port (int): The port number associated with the `neighbor_iface`.
```
Comment (yxieca) [inline tests/drop_counters/mock_server.py:28]: This is not really neighbor_port, this is port index on ptf. The naming of this variable is a bit confusing.

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    Gathers parameters about the testbed for the test cases to use.
+
+    Returns: A Dictionary with the following information:
+        "ports": A List of the active ports in the system
+        "phy_intfs": A List of ALL physical interfaces in the system
+        "vlan_intfs": A Dictionary containing all the VLAN interfaces in the system, including
+            their address, subnet, prefix length, and VLAN members.
+    """
+
+    if testbed["topo"]["name"] != "t0":
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:67]: You probably should check 'type' instead of 'name'. If I am not mistaking, name would be t0-64, t0-116, etc. type would be 't0'?

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    Returns:
+        A Dictionary of device capabilities (see `get_device_capabilities` under the
+        `configurable_drop_counters` package).
+    """
+
+    capabilities = cdc.get_device_capabilities(duthost)
+
+    if not capabilities:
+        pytest.fail("Error fetching device capabilities")
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:105]: condense to pytest_assert?

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        duthost.command("sonic-clear dropcounters")
+
+        ptf_tx_port_id = testbed_params["phy_intfs"].index(rx_port)
+        _send_packets(duthost, ptfadapter, pkt, ptf_tx_port_id)
+        time.sleep(3)
+
+        recv_count = cdc.get_drop_counts(duthost, counter_type, "TEST", rx_port)
+
+        pytest_assert(recv_count == PACKET_COUNT,
+                      "Expected {} drops, received {}".format(PACKET_COUNT, recv_count))
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:164]: Is there a chance to use wait_until here to get rid of the time.sleep above?

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    """
+    Mocks the presence of a server beneath a T0.
+
+    Returns:
+        A MockServer which will allow the caller to mock the behavior of a server within
+        a VLAN under a T0.
+    """
+
+    if not testbed_params["vlan_intfs"]:
+        pytest.skip("No VLANs available to mock a server under T0")
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:179]: Likely not belong to this PR. I am thinking we could use a pytest_requires helper (similar to pytest_assert).

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+        recv_count = cdc.get_drop_counts(duthost, counter_type, "TEST", rx_port)
+
+        pytest_assert(recv_count == PACKET_COUNT,
+                      "Expected {} drops, received {}".format(PACKET_COUNT, recv_count))
+
+    return _runner
+
+@pytest.fixture
+def mock_server(ptfhost, fanouthosts, testbed_params):
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:169]: I feel that this fixture can be shared by other tests too. Making it more generic?

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        pytest.skip("No VLANs available to mock a server under T0")
+
+    # Randomly pick one of the VLANs beneath the T0.
+    vlan = testbed_params["vlan_intfs"][random.choice(list(testbed_params["vlan_intfs"]))]
+
+    # Generate a dummy IP that falls under the VLAN subnet. Make sure it isn't
+    # the same as the VLAN interface IP!
+    server_ip = vlan["addr"]
+    while server_ip == vlan["addr"]:
+        server_ip = random.choice(list(IPNetwork(vlan["subnet"])))
```
Comment (yxieca) [inline tests/drop_counters/test_configurable_drop_counters.py:188]: Do you care if 2 servers randomized to the same IP?

File: tests/drop_counters/test_configurable_drop_counters.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    yield _setup_counters
+
+    try:
+        cdc.delete_drop_counter(duthost, "TEST")
+        time.sleep(1)
+    except Exception:
+        logging.info("Drop counter does not exist, skipping delete step...")
+
+@pytest.fixture
+def test_runner(duthost, ptfadapter, testbed_params):
```
Comment (wangxin) [inline tests/drop_counters/test_configurable_drop_counters.py:145]: Name `test_runner` looks like a test case. Maybe we should avoid using names with prefix `test_` or suffix `_test` if it is not a test case?

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 9 above)
Comment (wangxin) [reply tests/drop_counters/test_configurable_drop_counters.py:169]: Maybe we can consider using the arp_responder too. The arp_responder can automatically respond to ARP requests according to configuration. The configuration specifies which PTF interface should respond to ARP queries for what IP addresses.

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 10 above)
Comment (wangxin) [reply tests/drop_counters/test_configurable_drop_counters.py:188]: Yes, there is still possibility that the new randomly chosen IP is same as the VLAN interface IP. This kind of random failure would be hard to debug.

File: tests/drop_counters/mock_server.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
+"""
+Utilities for mocking a server using PTF.
+"""
+
+import logging
+import time
+
+from common.platform.device_utils import fanout_switch_port_lookup
+
+class MockServer(object):
```
Comment (tahmed-dev) [inline tests/drop_counters/mock_server.py:10]: Do you have plans to extend the functionality here? The reason I ask is that the name MockServer is a bit generic. Also, PtfProxy as a name as it appear to funnel/proxy commands to ptf host.

File: tests/drop_counters/configurable_drop_counters.py (hunk 3 above)
Comment (daall) [reply tests/drop_counters/configurable_drop_counters.py:148]: Part of it is because some of the offsets are for the rows and some of them are for the columns, so they need to be handled in different places.

I kept both of the row offsets because I thought it was more clear since logically we:
1. Skip any leading whitespace
2. Get the names of the counters from the first row of "real" output
3. Skip ahead to the actual counts

We could just fetch the counter names and then delete everything before the counts, but I'm not sure if this improves the performance or readability.

File: tests/drop_counters/configurable_drop_counters.py (hunk 2 above)
Comment (daall) [reply tests/drop_counters/configurable_drop_counters.py:75]: The output of the show command looks sorta like this:
```
COUNTER_TYPE:
    REASON
    REASON
COUNTER_TYPE:
   REASON
```

So if the line is not one of the counter type keywords, then it will be a drop reason.

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 9 above)
Comment (daall) [reply tests/drop_counters/test_configurable_drop_counters.py:168]: Thanks Xin, I'll take a look at arp_responder!

File: tests/drop_counters/mock_server.py (hunk 12 above)
Comment (daall) [reply tests/drop_counters/mock_server.py:10]: I agree. I think this might be good to have in the future, but ArpResponder may suffice for what we're trying to do here. If we do end up keeping this then I'll probably update the name and/or move this to common.

**Comment by daall** [comment]
retest this please

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 9 above)
Comment (daall) [reply tests/drop_counters/test_configurable_drop_counters.py:168]: I was finally able to get arp responder set up for my testbed and this is actually exactly what I was looking for in the first place! I will go ahead and refactor my test to use arp_responder and get rid of MockServer, since arp_responder is doing the same thing in a much more robust way.

File: tests/common/devices.py [hunk 13]
```diff
@@ -553,6 +553,19 @@ def get_version(self):
...
         return output["stdout_lines"][0].strip()
+    
+    def get_dut_iface_mac(self, iface_name):
+        """
+        Gets the MAC address of specified interface.
+
+        Returns:
+            str: The MAC address of the specified interface, or None if it is not found.
+        """
+        for iface, iface_info in self.setup()['ansible_facts'].items():
```
[the remark this answers is not in the document]
Comment (daall) [reply tests/common/devices.py:607]: I think that's a good idea but I'm thinking I will save that for a follow-up PR since the change also needed to be vetted for the existing `test_drop_counters.py` script.

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 10 above)
Comment (daall) [reply tests/drop_counters/test_configurable_drop_counters.py]: Fixed!

File: tests/drop_counters/test_configurable_drop_counters.py (hunk 9 above)
Comment (daall) [reply tests/drop_counters/test_configurable_drop_counters.py]: And done!

**Comment by daall** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please

**Comment by daall** [comment]
retest vsimage please
